controllers/wicd_csr_controller.go

```diff
@@ -0,0 +1,167 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/openshift/windows-machine-config-operator/pkg/rbac"
+	"github.com/openshift/windows-machine-config-operator/pkg/windows"
+	certificatesv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8sapierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/openshift/windows-machine-config-operator/pkg/cluster"
+	csrvalidation "github.com/openshift/windows-machine-config-operator/pkg/csr/validation"
+)
+
+//+kubebuilder:rbac:groups="certificates.k8s.io",resources=certificatesigningrequests/approval,verbs=update
+//+kubebuilder:rbac:groups="certificates.k8s.io",resources=certificatesigningrequests,verbs=get;list;watch
+//+kubebuilder:rbac:groups="certificates.k8s.io",resources=signers,verbs=approve,resourceNames=kubernetes.io/kube-apiserver-client
+//+kubebuilder:rbac:groups=certificates.k8s.io,resources=certificatesigningrequests/status,verbs=get;update;patch
+
+const (
+	// WICDCSRController is the name of this controller in logs
+	WICDCSRController = "wicd-csr"
+)
+
+// wicdCSRReconciler handles certificate signing requests for WICD
+type wicdCSRReconciler struct {
+	instanceReconciler
+	validator *csrvalidation.CSRValidator
+}
+
+// NewWICDCSRController creates a new WICD CSR controller following the existing pattern
+func NewWICDCSRController(mgr manager.Manager, clusterConfig cluster.Config, watchNamespace string) (*wicdCSRReconciler, error) {
+	clientset, err := kubernetes.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		return nil, fmt.Errorf("error creating kubernetes clientset: %w", err)
+	}
+
+	validator := csrvalidation.NewCSRValidator(mgr.GetClient(), csrvalidation.WICDCertType)
+
+	return &wicdCSRReconciler{
+		instanceReconciler: instanceReconciler{
+			client:             mgr.GetClient(),
+			log:                ctrl.Log.WithName("controllers").WithName(WICDCSRController),
+			k8sclientset:       clientset,
+			clusterServiceCIDR: clusterConfig.Network().GetServiceCIDR(),
+			watchNamespace:     watchNamespace,
+			recorder:           mgr.GetEventRecorderFor(WICDCSRController),
+		},
+		validator: validator,
+	}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager
+func (r *wicdCSRReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Build controller for CertificateSigningRequests
+	return builder.ControllerManagedBy(mgr).
+		For(&certificatesv1.CertificateSigningRequest{}).
```

Comment:
Any possibility of predicates here, so that we don't call reconcile on all CSR events in the cluster? I think at the least you need to filter based on event type.